{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The pipeline to analyse Asteraceae genomes using Asteraceae synteny-phylogenomic framework"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get dependencies ready\n",
    "import os\n",
    "from pathlib import Path\n",
    "from utils import processLCSAndFirstFilter_2sp as plff\n",
    "from utils import processFinalFilter_2sp as pff\n",
    "import shutil\n",
    "from processDrimm_tao_2sp import readSequence,prepare_dir,write_raw_blocks,processDrimm\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from modules_paper import prepare_OG,prepare_meta,make_comBed,combed_parser,grimmBlock_parser,getFilterSequence,getAllSequence,outSequence,standarlize_drimmblocks,get_block_info,get_chromosome_info\n",
    "from parse_genespaceBED_for_drimmBED_paper import make_index,make_drimmBED,make_genome_meta,make_chromosome_meta,get_chromosome_lenght,parse_genespaceBED_for_drimmBED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get input data ready\n",
    "sp='sp46'\n",
    "index_file = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/Smar.EM05.v1.genome.fa.fai\"\n",
    "\n",
    "workdir = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/workdir\"\n",
    "BED_folder = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/\"\n",
    "if not workdir[-1] == \"/\": workdir += \"/\"\n",
    "if not BED_folder[-1] == \"/\": BED_folder += \"/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step0, first let's prepare bed and meta data\n",
    "## make index file\n",
    "df_index = make_index(sp,index_file) #your genome index and genome name you want to use\n",
    "\n",
    "## make bed, genome_meta, chromosome_meta\n",
    "DIC = {f\"{row['sp']}@{row['chr']}\": row['length'] for _, row in df_index.iterrows()}\n",
    "bed_dir = workdir + \"bed/\"\n",
    "if not os.path.exists(bed_dir):\n",
    "    os.makedirs(bed_dir)\n",
    "    make_drimmBED(BED_folder, DIC, bed_dir)\n",
    "else:\n",
    "    print(bed_dir + \" exists, let's generate bed file in this floder\")\n",
    "    make_drimmBED(BED_folder, DIC, bed_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step1 process synOG\n",
    "## get dependent info ready, read in synOG table\n",
    "from processOrthofinder_tao import get_group_dir,get_final_group,main\n",
    "\n",
    "## read in synOG table and adjust name\n",
    "orthogroups = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/2_analysis/1_genespace/run30/pangenome_sp03v5_synOG_2sp_syn.txt\"\n",
    "## be careful with this orthogroups table; in terms of the name \"sp03v5\" vs \"\"sp03_v5\"\", and \"sp03v5@gene1\" vs \"sp03v5_gene1\"; we need to make this consistent in genespace_parser()\n",
    "ortho = pd.read_csv(orthogroups, sep='\\t', dtype = str, keep_default_na=False)\n",
    "ortho2 = ortho[ortho[\"interpChr\"] != \"\"]\n",
    "\n",
    "ortho2.rename(columns={'sp03_v5': 'sp03v5', 'sp03_v1': 'sp03v1'}, inplace=True)\n",
    "ortho2['sp03v5'] = ortho2['sp03v5'].str.replace('sp03_v5@', 'sp03v5@')\n",
    "ortho2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## process synOG table, call pairwise orthologs for given ref-target pair\n",
    "##target = [\"sp28\",\"sp24\",\"sp43\",\"sp44\",\"sp18\",\"sp34\",\"sp14\",\"sp10\",\"sp08\",\"sp35\",\"sp05a\",\"sp39\",\"sp37\",\"sp11\",\"sp13\",\"sp22\",\"sp21\",\"sp29\",\"sp30\",\"sp03_v5\",\"sp03_v1\",\"sp03\",\"sp41\",\"sp07\",\"sp15\"]\n",
    "target = [\"sp46\"]\n",
    "reference = \"sp03v5\"\n",
    "OG_type = \"synOG\" # always check the varable \"sp_ratio\"\n",
    "\n",
    "genome_meta = make_genome_meta(\"sp46\",1,17)\n",
    "chromosome_meta = make_chromosome_meta(BED_folder,DIC)\n",
    "print(chromosome_meta)\n",
    "\n",
    "outdir = workdir + \"out1\"\n",
    "if outdir[-1] != \"/\": outdir += \"/\"\n",
    "if bed_dir[-1] != \"/\": bed_dir += \"/\"\n",
    "\n",
    "for species in target:\n",
    "    if not species == reference: \n",
    "        species_list = [reference, species]\n",
    "        subdir = outdir + species + \"/1_SynOG/\"\n",
    "        if not Path(subdir).exists():\n",
    "            os.makedirs(subdir)\n",
    "            sp,sp_ratio,sp_chr_number,long_chr_list,gff_list = prepare_meta(species_list,genome_meta,chromosome_meta,OG_type,bed_dir)\n",
    "            #print(sp)\n",
    "            #print(sp_chr_number)\n",
    "            print(gff_list,long_chr_list)\n",
    "            ortho3 = prepare_OG(ortho2,sp,OG_type)\n",
    "            group_dir = get_group_dir(ortho3,sp)\n",
    "            finalGroup = get_final_group(group_dir,sp_ratio)\n",
    "            main(bed_dir,subdir,group_dir,finalGroup,gff_list,long_chr_list)\n",
    "        else: print(species + \" seems already done, please check!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "bat"
    }
   },
   "source": [
    "### Step2 perform drimm-synteny in windows\n",
    "1. transfer the data from last step to windows\n",
    "   #put the whole directory (e.g., genespace26) in C:\\Users\\fengtao\\Desktop\\drimm_batch\\\n",
    "2. double click the windows bat file \"drimm-synteny\"; the tool will loop the genomes in \"genespace26\" and perform calculation\n",
    "   \n",
    "   **input poped options:**\n",
    "     + The path of drimm.sequence <drimm.sequence>\n",
    "     + The output directory <.>\n",
    "     + cycleLengthThreshold controls the continuity of synteny blocks (default parameter is 20)\n",
    "     + dustThreshold controls the upper limit of gene family. The gene family will be filtered when homologous genes exceeding dustThreshold. For above example, target copy numbers are 2,4,2,2,2 in each species, then the dustThreshold is 13 (2+4+2+2+2+1)\n",
    "3. wait untill the calculations done, copy the results to MAC, and go ahead with next step\n",
    "\n",
    "A  ***drimm-synteny.bat*** file is like this:\n",
    ">\n",
    "> @echo off\n",
    ">\n",
    "> REM Loop through each subdirectory\n",
    "> for /d %%A in (\"C:\\Users\\fengtao\\Desktop\\drimm_batch\\genespace26\\*\") do (\n",
    "   >\n",
    "    > echo Entering directory: %%A\\2_DrimmRaw\\\n",
    "    >\n",
    "    > mkdir \"%%A\\2_DrimmRaw\\\"\n",
    "    >\n",
    "    > copy \"%%A\\1_SynOG\\drimm.sequence\" \"%%A\\2_DrimmRaw\\\"\n",
    "    >\n",
    "    > pushd \"%%A\\2_DrimmRaw\\\"\n",
    "    >\n",
    "    >\n",
    "    > echo inputs like this, modify accordingly: inputfile=drimm.sequence  outdir=.  cycleLengthThreshold=20  dustThreshold=5\n",
    "    >\n",
    "    >\n",
    "    > REM Run your executable tool with non-interactive inputs\n",
    "    >\n",
    "    > \"C:\\Users\\fengtao\\Desktop\\processDrimm\\drimm\\drimm\\bin\\Release\\netcoreapp3.1\\win-x64\\drimm.exe\"\n",
    "    >\n",
    "    > echo %%A done!\n",
    "    >\n",
    "    > popd\n",
    "> )\n",
    "> pause"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# step3 process drimm\n",
    "# from drimm-synteny, we get raw blocks, and we need to clean and sort the raw blcoks to\n",
    "## 1. split the all-in-one blocks (drimm-synteny output) into individual genomes\n",
    "## 2. meanwhile divide the blocks into separate ones based on ratio (normally from genome duplication, but segmental duplications cannot be distinguished, and are also processed)\n",
    "workdir = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/workdir\"\n",
    "BED_folder = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/\"\n",
    "if workdir[-1] != \"/\": workdir += \"/\"\n",
    "if BED_folder[-1] != \"/\": BED_folder += \"/\"\n",
    "\n",
    "outdir = workdir + \"out1/\"\n",
    "\n",
    "sp='sp46'\n",
    "index_file = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/Smar.EM05.v1.genome.fa.fai\"\n",
    "df_index = make_index(sp,index_file) #your genome index and genome name you want to use\n",
    "DIC = {f\"{row['sp']}@{row['chr']}\": row['length'] for _, row in df_index.iterrows()}\n",
    "\n",
    "genome_meta = make_genome_meta(\"sp46\",1,17)\n",
    "chromosome_meta = make_chromosome_meta(BED_folder,DIC)\n",
    "\n",
    "#bed_path = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/2_analysis/10_ancestralGenome/bed\"\n",
    "OG_type = \"synOG\" # synOG|orthofinder\n",
    "working_dir = outdir\n",
    "bed_dir = workdir + \"bed/\"\n",
    "\n",
    "if working_dir[-1] != \"/\": working_dir += \"/\"\n",
    "if bed_dir[-1] != \"/\": bed_dir += \"/\"\n",
    "\n",
    "target = [\"sp46\"]\n",
    "reference = \"sp03v5\"\n",
    "\n",
    "for species in target:\n",
    "    if not species == reference: \n",
    "        species_list = [reference, species]\n",
    "        sp_working_dir = working_dir + species + \"/\"\n",
    "        print(sp_working_dir)\n",
    "        homolog_dir = sp_working_dir + \"1_SynOG/\"\n",
    "        if not Path(sp_working_dir).exists():\n",
    "            print(\"no homologs found for \" + species)\n",
    "        else:\n",
    "            block_file = sp_working_dir + \"/2_DrimmRaw/blocks.txt\"\n",
    "            sequence = readSequence(block_file)\n",
    "            drimmSyntenyFile = sp_working_dir + \"/2_DrimmRaw/synteny.txt\"\n",
    "            \n",
    "            list_ = prepare_meta(species_list, genome_meta, chromosome_meta, OG_type, bed_dir)\n",
    "            sp_list = list_[0]\n",
    "            sp_ratio = list_[1]\n",
    "            sp_ratio = \":\".join([str(e) for e in sp_ratio]) # convert sp_ratio to a string\n",
    "            chr_number = list_[2]\n",
    "            \n",
    "            drimm_split_blocks_dir,raw_block_dir,result_dir = prepare_dir(sp_working_dir)\n",
    "            write_raw_blocks(sequence,chr_number,drimm_split_blocks_dir,sp_list)\n",
    "            processLCSAndFirstFilter = plff.processLCSAndFirstFilter(drimm_split_blocks_dir, raw_block_dir, sp_ratio,\n",
    "                                                         drimm_split_blocks_dir, homolog_dir, drimmSyntenyFile,\n",
    "                                                         sp_list, 's')\n",
    "            processLCSAndFirstFilter.excute()\n",
    "\n",
    "            processFinalFilter = pff.processFinalFilter(sp_list, raw_block_dir, drimm_split_blocks_dir, result_dir, 's')\n",
    "            processFinalFilter.excute()\n",
    "            shutil.rmtree(raw_block_dir)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use grimm2barplot to to do chromosome painting\n",
    "## use R packages chromoMap to paint blocks on full chromosme background (a difference to iage_plot function);\n",
    "## 2.1 let's first make a combed file based on normal bed, this special bed will be used by combed_parser() and grimmBlock_parser()\n",
    "beddir = BED_folder\n",
    "comBed = make_comBed(beddir)\n",
    "#comBed.to_csv(workdir + 'combBed.txt', sep='\\t', header=True, index=False)\n",
    "\n",
    "## 2.2 let's read in cleaned dirmm blocks, parse it and generate new data <5_SynBlocks>; no need to worry about ratio as all scenarios will be considered in one run\n",
    "working_dir = workdir + \"out1\"\n",
    "if not working_dir.endswith(\"/\"): working_dir += \"/\"\n",
    "\n",
    "target = [\"sp46\"]\n",
    "\n",
    "for species in target:\n",
    "    blockDir = working_dir + species + \"/3_DrimmBlocks/finalBlocks/\"\n",
    "    output_path = blockDir + \"../../5_SynBlocks/\"\n",
    "    if not os.path.exists(output_path): os.mkdir(output_path)\n",
    "    dic = combed_parser(comBed)\n",
    "    blockFiles = [os.path.join(blockDir, filename) for filename in os.listdir(blockDir) if filename.endswith('.synteny.genename')]\n",
    "    grimmBlock_parser(blockFiles,dic,output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 2.3 take outputs from last step, generate table for plotting in R package <chromoMap>; no need to worry about ratio as all scenarios will be considered in one run\n",
    "\n",
    "target = [\"sp46\"]\n",
    "reference = \"sp03v5\"\n",
    "working_dir = workdir + \"out1\"\n",
    "if not working_dir.endswith(\"/\"): working_dir += \"/\"\n",
    "\n",
    "for species in target:\n",
    "    blockDir = working_dir + species + \"/5_SynBlocks/\"\n",
    "    blocks = [file for file in os.listdir(blockDir) if file.endswith(\"synBlock.txt\")]\n",
    "    for block in blocks:\n",
    "        target_block_name = block.split(\"_\")[0]\n",
    "        target_block_type = block.split(\"_\")[1]\n",
    "        reference_block = blockDir + reference + \"_\" + target_block_type + \"_synBlock.txt\"\n",
    "        target_block = os.path.join(blockDir, block)\n",
    "        target_chr = blockDir + \"../3_DrimmBlocks/finalBlocks/\" + target_block_name + \"_\" + target_block_type +\".final.block\"\n",
    "        \n",
    "        dic1,dic2 = get_block_info(reference_block,target_block)\n",
    "        coo_gene,coo_bp = get_chromosome_info(target_chr,dic1,dic2)\n",
    "\n",
    "        columns = [\"chr\",\"haplotype\", \"start\", \"end\", \"ancestral_group\"]\n",
    "        output_path = blockDir + \"../6_chromPainting/\"\n",
    "        if not os.path.exists(output_path): os.mkdir(output_path)\n",
    "\n",
    "        r1 = pd.DataFrame(coo_gene, columns=columns) # make a dataframe from the list of list\n",
    "        r2 = pd.DataFrame(coo_bp, columns=columns) # make a dataframe from the list of list\n",
    "        out_put1 = output_path + target_block_name + \"_\" + target_block_type +\"_cooGene.txt\"\n",
    "        out_put2 = output_path + target_block_name + \"_\" + target_block_type +\"_cooBp.txt\"\n",
    "        r1.to_csv(out_put1, sep='\\t', header=True, index=False)\n",
    "        r2.to_csv(out_put2, sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 2.4 generate inputs for chromoMap\n",
    "sp='sp46'\n",
    "index_file = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/Smar.EM05.v1.genome.fa.fai\"\n",
    "df_index = make_index(sp,index_file) #your genome index and genome name you want to use\n",
    "DIC = {f\"{row['sp']}@{row['chr']}\": row['length'] for _, row in df_index.iterrows()}\n",
    "\n",
    "genome_meta = make_genome_meta(\"sp46\",1,17)\n",
    "chromosome_meta = make_chromosome_meta(BED_folder,DIC)\n",
    "\n",
    "target = [\"sp46\"]\n",
    "\n",
    "workdir = \"/Users/fengtao/Library/CloudStorage/OneDrive-WageningenUniversity&Research/project/Asteraceae_evolution/Asteraceae_phylogenomics/pipeline/workdir\"\n",
    "working_dir = workdir + \"/out1\"\n",
    "if working_dir[-1] != \"/\": working_dir += \"/\"\n",
    "\n",
    "# obtain information of chromosome number in all species from genome_meta\n",
    "#genome_meta_df = pd.read_csv(genome_meta, sep='\\t', dtype = str, keep_default_na=False, names=['genome', 'ratio1', 'ratio2','chromosome'])\n",
    "genome_meta_dict = genome_meta.set_index('sp')['chrN'].to_dict()\n",
    "chromosome_meta_dict = {lst[0]: lst[1:] for lst in chromosome_meta}\n",
    "\n",
    "for genome in target:\n",
    "    outdir = working_dir + genome + \"/\"\n",
    "    chromosome_list = chromosome_meta_dict[genome]\n",
    "    chr_info_list = []\n",
    "    i = 1\n",
    "    for chromosome in chromosome_list:\n",
    "        if i <= int(genome_meta_dict[genome]): # make sure only take the long chromosomes as specified in genome_meta\n",
    "            drimm_chr = \"chr_\" + chromosome.split(\"_\")[-2]\n",
    "            genome_chr = \"_\".join(chromosome.split(\"_\")[0:-2])\n",
    "            length = chromosome.split(\"_\")[-1]\n",
    "            chr_info = [drimm_chr,genome_chr,\"0\",length]\n",
    "            chr_info_list.append(chr_info)\n",
    "            i += 1\n",
    "    df = pd.DataFrame(chr_info_list, columns=['drimm_chr', 'genome_chr', 'start','end'])\n",
    "    df = df.sort_values(by='genome_chr')\n",
    "    df.to_csv(outdir+genome+\".txt\",sep=\"\\t\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
